### ISSUE TITLE
nfsp_agent samples best-response instead of average policy

### ISSUE THREAD

**Issue by mjudell** [datamllab/rlcard#50]
It looks like nfsp_agent samples the best-response network in [evaluation mode](https://github.com/datamllab/rlcard/blob/3456ebc04cafc399417276b15168cdcb809752f3/rlcard/agents/nfsp_agent.py#L185).  I copied this behavior in the PyTorch [implementation](https://github.com/mjudell/rlcard/blob/861270babf08df1128dafb3072d2b9a29970fb38/rlcard/agents/nfsp_agent_pytorch.py#L175).  However, Theorem 7 in [1] argues that it is the average strategy profile that converges to a Nash equillibrium.  Sampling the best-response network produces a deterministic pure strategy, while the average policy network produces a stochastic behavioural strategy.  This is discussed in Section 4.2 of [2].  Also, it looks like DeepMind's implementation [3] samples the average policy network in evaluation [mode](https://github.com/deepmind/open_spiel/blob/8b2501b49877dadcce26912f62b46024e17e56f3/open_spiel/python/examples/kuhn_nfsp.py#L101).

Am I missing something? 

References:
[1] Heinrich et al. (2015) "Fictitious Self-Play in Extensive-Form Games"
[2] Heinrich and Silver (2016) "Reep Reinforcement Learning from Self-Play in Imperfect Information Games"
[3] Lanctot et al. (2019) "OpenSpiel: A Framework for Reinforcement Learning in Games"

**Comment by daochenzha** [comment]
@mjudell Thank you for pointing out. And thank you for the efforts for PyTorch implementation. We are currently reviewing your implementation.

Yes. It is the average policy rather than best response that converges to Nash Equilibrium.

The current code uses best response because we found that best response performs equally well (or seems even better) in our preliminary experiments. Thus, we leave best response there. It may be caused by suboptimal hyperparameters, the network structure, or the evaluation metrics since we currently use tournaments and human-interfaces playing experiences for evaluation which may be biased. We will soon implement exploitation metric, which should be better.

From a scientific view, we should definitely use average policy as in the original paper. We will correct the codes and do more experiments with average policy and exploitation metric. Thanks:)

**Comment by daochenzha** [comment]
@mjudell Section 4.2 in [2] shows that self-play deterministic, greedy strategy (such as DQN) will lead to highly exploitable policy. It is also confirmed in our experiments. We found that although DQN performs well against random agents, it looks worse than NFSP when they play against each other.

It seems that the best-response policy in NFSP also delivers reasonably good performance (better than DQN). It makes sense to me since the best-response policy is mixed with the slowly changed average policy in training.

Correct me if my understanding is not correct :)

**Comment by mjudell** [comment]
Sounds right to me, and it looks like Figure 2(a) in [2] also finds that the best response outperforms the average policy against SmooCT.

**Comment by daochenzha** [comment]
@mjudell Thanks for letting me know. I didn’t notice this figure. It seems to be better if using an argument to decide whether average policy or best response should be used for evaluation since they both seem to work.
